llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp

```diff
@@ -131,6 +131,13 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
   Register SrcReg = Src.getReg();
 
   if (isVCC(DstReg, *MRI)) {
+    // Allow copy from physical register other than SCC to s1.
+    if (SrcReg.isPhysical() && SrcReg != AMDGPU::SCC) {
+      const TargetRegisterClass *DstRC = MRI->getRegClassOrNull(DstReg);
```

Comment:
Shouldn't be re-querying the class of DstReg, isVCC should tell you everything you need to know here

Reply:
ping

Reply:
If we don't get the reg class of DstReg, how to check if it's compatible with the src physical reg?

Reply:
isVCC should have considered that in the first place. isVCC should already be telling you the class

Reply:
I'm not sure I follow. `isVCC()` here checks on `DstReg` to see if its virtual and if its size=1. After that's confirmed, the new code checks if a physical reg can be copied to it. For example, on wave64, copying $sgpr4_sgpr5 is ok but not $sgpr4. Note that the new code is similar to the existing code immediately below that checks for `SrcReg = SCC`.

Reply:
You have to solve exactly the same problem as you did in the RegBankSelect pass. In the physical register case, you need to consider the virtual register type from the other operand. Querying the class on the destination register after checking isVCC does not give you additional information, isVCC needs the additional context for the physical register case

Reply:
Are you suggesting modifying `isVCC()` and also passing the src register as an argument?

Reply:
@arsenm Ping.

Reply:
Yes. You need the same set of information as your getInstrMapping patch

Reply:
After another look, I think the problem here is quite different from the one in `getInstrMapping()`. Here we are dealing with instruction selection for a COPY from a physical reg to a virtual reg of type s1, such as `%0:vcc(s1) = COPY $sgpr0`. The existing `selectCOPY()` function would reject it. This patch is trying to allow such COPY in some cases. It takes advantage of an existing check of `isVCC()` on the dst reg, and then handles the case when the src is a physical reg. Then patch simply makes sure the src physical reg belongs in the reg class of the dst reg. For example, the above instruction would be allowed on a wave32 target but not on a wave64 target.

In `getInstrMapping()` the reg bank is not selected yet, so based on the type being s1 we set the reg bank to VCCRegBank to allow the mappting to succeed. Here, reg bank is already determined, and `isVCC()` already checks to make sure the type is s1. So I'm not sure what more info should be given to `isVCC()`.

Reply:
> After another look, I think the problem here is quite different from the one in getInstrMapping()

The situation here is isomorphic. A physical register alone is not enough context to determine if it is interpreted as a boolean or not. You need the additional hint from the virtual register in the other operand to determine that, just like in the regbankselect case.